### ISSUE TITLE
Intermittent 503 Errors with HTTP2 Client

### ISSUE THREAD

**Issue by dev-chirag** [envoyproxy/envoy#20650]
*Description*:
I am running a load test with Envoy as load balancer and an HTTP2 client. The test is mostly stable, but intermittently I see a lot of 503 errors on the client. For the same requests, I can see that my backend server sends a 200 to Envoy. 

In the TCP dump I see 503 errors with an error - **[upstream connect error or disconnect/reset before headers. reset reason: remote refused stream**

*Repro steps*:

*Admin and Stats Output*:

*Config*:

*Logs*:
I couldnt find anything error related in the Envoy logs

*Call Stack*:

**Comment by lambdai** [comment]
IIUC, when Envoy reports `upstream connect error or disconnect/reset before headers. reset reason: remote refused stream`, Envoy won't encode the request to backend *unless Envoy does retry*, there is no chance to see the backend send 200 to Envoy. 

Can you confirm Envoy-Backend 200 and client-Envoy-503 are the corresponding request? I believe the trace id can be used to identify the downstream upstream request

**Comment by dev-chirag** [comment]
@lambdai you are correct. My bad, I messed up my tracing in Wireshark. I can see that the backend sent a RST_STREAM (REFUSED_STREAM) on which Envoy sends a 503, which is acceptable.

Now this maybe a wild guess but I am wondering why the backend will send a REFUSED_STREAM. Could [this](https://datatracker.ietf.org/doc/html/rfc7540#section-5.1.2) be the reason for this? 

Does Envoy honor the MAX_CONCURRENT_STREAMS settings from the server?

**Comment by lambdai** [comment]
Thank you for the confirmation.

The answer should be yes but I also find this comment:

https://github.com/envoyproxy/envoy/blob/main/source/common/http/conn_pool_base.cc#L113-L116
```
// Note: if multiple streams are assigned to a connection before the settings
// are received, they may still be reset by the peer. This could be avoided by
// not considering http/2 connections connected until the SETTINGS frame is
// received, but that would result in a latency penalty instead.
```
Since you are running load testing, it's possible you are hitting it.

CC @alyssawilk

**Comment by alyssawilk** [comment]
if you know what the setting is going to be, you can configure concurrent stream limits for upstream H2 to not exceed the cap.  Otherwise I think you'd just have to configure retries.  @diannahu was just asking about Envoy retry behavior for going over the stream limits and it'd be kinda cool if we auto-retried like @danzh2010 is doing for various early HTTP/3 behaviors, but that's not implemented. 

Also I was planning on picking up https://github.com/envoyproxy/envoy/issues/20696 later this month and that would help after the first time talking to an endpoint.

**Comment by egerkke** [comment]
Hi Lady and Gents,

Some further clarification:
 - the executed load test is also a long (12 hours) stability test. Therefore it is hard to catch the issue with tcpdump when it is happening
 - our upstream server as a peer is limiting the MAX_CONCURRENT_STREAMS to 70 in SETTING frame (we could verify this in tcpdump)

My suspicion is that from time to time as the upstream is loaded we start reaching the configured limit and when overflow is happening then there is a protocol error happening that results in HTTP 503 on client side.

Since it is hard to capture the issue, I used some tools this ticket:
    https://github.com/envoyproxy/envoy/issues/13933
Downloaded referred go program (that propagates its MAX_CONCURRENT_STREAMS as 5):
    https://gist.github.com/kryzthov-stripe/8d8b5f5f1bd7cb99a88baba08f1c5058
Created a simple envoy configuration file:
    base_envoy_config.txt

I attached my source code as max_stream_problem.tgz

To execute the test, start the server:
    go run server.go  //you might need to get modules by executing "go get golang.org/x/net/http2/h2c" and "go get golang.org/x/net/http2"
Start Envoy:
    docker run -it --network host --mount type=bind,source=<path_to_config_file>/base_envoy_config.txt,target=/config.yaml envoyproxy/envoy-dev:latest /usr/local/bin/envoy -c config.yaml --service-cluster front-proxy --concurrency 1
Execute hey:
    go get github.com/rakyll/hey
    hey -c 200 -t 0 http://localhost:8000

In case the concurrency is set for 1, the result was
Status code distribution:
  [200]	10 responses
  [503]	190 responses

Also please note that one of our 3PP Contour does the Envoy configuration and it does not support currently to set MAX_CONCURRENT_STREAMS parameter for upstream clusters. Nevertheless according to https://datatracker.ietf.org/doc/html/rfc7540 my understanding is that we as peer shall be able to inform Envoy about our limitation via the SETTINGS frame.

May I ask your valuable help in clarifying this if my suspicion is right? Many thanks in advance!
[max_stream_problem.tar.gz](https://github.com/envoyproxy/envoy/files/8543884/max_stream_problem.tar.gz)

**Comment by alyssawilk** [comment]
Yeah if tweaking concurrent streams fixes the problem, it's likely a disagreement about the stream limits which is the root cause.
You can inform Envoy of stream limits via the settings frame and AFIK Envoy respects that, but if you have 80 streams queue, no upstream stream limits configured, and establish a new connection, it'll happily dump 80 streams on there before it receives the SETTINGS limiting streams to 70.  https://github.com/envoyproxy/envoy/issues/20696 should improve by caching once it lands, but if you cant set the MAX_CONCURRENT_STREAMS there's not much else you can do today to address the inherent race condition.

**Comment by egerkke** [comment]
Hej Lady,
Thanks for the answer! Is it possible to somehow catch in our case if this is the issue that new connection is established, but there is already 70+ streams waiting to be sent via the connection? Can we see this somewhere? Or only with Envoy debug log?
Or the real test will be once [#20696](https://github.com/envoyproxy/envoy/issues/20696) is delivered and trying with this if it helps?

Thanks in advance!

**Comment by egerkke** [comment]
Hi Lady,

I see you delivered your solution:
Author: alyssawilk <alyssar@chromium.org>
Date:   Mon May 16 15:56:02 2022 -0400

    http2: using cached settings for multiplexing limits across the board (#21275)
    
    This turns up settings caching for HTTP/2 and ALPN upstreams, with each cluster caching the multiplexed streams for endpoints, and using those as the default for subsequent connections.
    
    Risk Level: high (pool capacity math is super complicated)
    Testing: prior integration testing, new unit tests
    Docs Changes: n/a
    Release Notes: inline
Is this planned to be included in Envoy 1.23.0?

**Comment by alyssawilk** [comment]
Yep!  Should make the next Envoy release (unless it causes problems of course)
